### ISSUE TITLE
encode/1 and decode/1 incompatibility

### ISSUE THREAD

**Issue by yegortimoshenko** [yegortimoshenko/json#5]
You can't encode a decoded JSON:

```
{error, _} = json:encode(json:decode(<<"{\"age\":25,\"children\":[\"Nick\",\"Travis\"],\"name\":\"Jim\"}">>)).
```

It doesn't work since in `json:decode/1` JSON strings are Erlang lists, while in `json:encode/1` they are Erlang binaries. It is to save decoding time. Probably I should address this issue.

**Comment by yegortimoschenko** [comment]
I've benchmarked `decode_string/2` with binary construction vs current `decode_string/2`. It seems that the former is 50% slower than the latter.

**Comment by yegortimoshenko** [comment]
Closed by 799963b2fe29a3d4dccdad6fead7a99589d81284.
